crucible-wasm/src/Lang/Crucible/Wasm/Memory.hs

```diff
@@ -0,0 +1,277 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE ImplicitParams #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Lang.Crucible.Wasm.Memory where
+
+import Data.Bits
+import qualified Data.BitVector.Sized as BV
+import qualified Data.ByteString.Lazy as LBS
+import Numeric.Natural
+import Data.Parameterized.Context
+
+import Lang.Crucible.Backend
+import Lang.Crucible.Simulator
+import Lang.Crucible.Simulator.Intrinsics
+import Lang.Crucible.Types
+
+import qualified Language.Wasm.Structure as Wasm
+
+import           Lang.Crucible.LLVM.Bytes
+import           Lang.Crucible.LLVM.DataLayout (EndianForm(..), noAlignment)
+import           Lang.Crucible.LLVM.MemModel
+import qualified Lang.Crucible.LLVM.MemModel.Generic as G
+import qualified Lang.Crucible.LLVM.MemModel.Partial as Partial
+
+import What4.Interface
+import What4.InterpretedFloatingPoint
+
+type WasmMem = IntrinsicType "Wasm_Mem" EmptyCtx
+
+-- | 64Ki, the number of bytes in a memory page, as defined
+--   by the WebAssembly spec.
+pageSize :: BV.BV 32
+pageSize = BV.word32 (64 * 1024)
+
+-- | 2^16 is the absolute maximum number of pages that can
+--   be allocated in a Wasm memory.
+maxPages :: BV.BV 32
+maxPages = BV.word32 (bit 16)
+
+-- | Implicit allocation number of the memory base pointer
+basePointer :: Natural
+basePointer = 1
+
+freshMemory :: IsSymInterface sym => sym -> Wasm.Limit -> IO (WasmMemImpl sym)
+freshMemory sym lim =
+  do (lo,hi) <- checkLim sym lim
+
+     -- allocate an entire 32-bit address space in little endian mode
+     let heap0 = G.emptyMem LittleEndian
+     let heap1 = G.allocMem @32 G.GlobalAlloc basePointer Nothing noAlignment G.Mutable "" heap0
+
+     -- clear the entire memory space by writing a constant array with the zero byte
+     z32  <- bvLit sym knownNat (BV.word32 0)
+     base <- natLit sym basePointer
+     let bp = LLVMPointer base z32
+
+     z8   <- bvLit sym knownNat (BV.word8 0)
+     zArr <- constantArray sym (Empty :> BaseBVRepr (knownNat @32)) z8
+     (heap,_,_) <- G.writeArrayMem sym (knownNat @32) bp noAlignment zArr Nothing heap1
+
+     pure WasmMemImpl
+          { wasmMemHeap = heap
+          , wasmMemSize = lo
+          , wasmMemMax  = hi
+          }
+
+checkLim :: IsSymInterface sym => sym -> Wasm.Limit -> IO (SymBV sym 32, SymBV sym 32)
+checkLim sym (Wasm.Limit lo Nothing)
+  | lo <= bit 16
+  = do lo' <- bvLit sym knownNat (BV.mkBV knownNat (fromIntegral lo))
+       hi' <- bvLit sym knownNat maxPages
+       pure (lo',hi')
+
+checkLim sym (Wasm.Limit lo (Just hi))
+  | lo <= hi && hi <= bit 16
+  = do lo' <- bvLit sym knownNat (BV.mkBV knownNat (fromIntegral lo))
+       hi' <- bvLit sym knownNat maxPages
```

Comment:
Should this be `hi` instead of `maxPages`?

Reply:
oh... good catch